Added support for creating a marketplace image in Alibaba

russellseymour opened the pull request:
Changes to the cookbook to allow Marketplace images to be created on the Alibaba platform.

As mentioned in https://github.com/chef-partners/omnibus-marketplace/pull/93 packages need to b uploaded to the OSS service so that they can be downloaded without the danger of time out. A number of changes were required to the cookbook to support this and the Alibaba cloud. This will mean that when new version of Omnibus Marketplace, Automate and Chef Server are created they will need to be uploaded to OSS before a package can be created.

The kitchen file for Alibaba states that Packer version 1.0.4 is used. This version has the Alibaba provider contained within it. It is also now one binary instead of lots of little files.

A new attribute file for Alibaba (`alibaba.rb`) has been created that holds the necessary configuration. It was in this file that I tried to pass the URL for the software packages that are held in OSS, but it would not pass them through to the `chef-marketplace` cookbook as attributes.

There is also an attribute set here that states the URL for the Omnibus Marketplace debian file. There is a companion template `install_marketplace_oss.sh.erb`, but although I tried to pass in the URL as a variable in `_publish.rb` using the `packer_provisioner` resource it would not render properly hence why it is hardcode at the moment. (I really do not now why this was the case and I will try it again).

The necessary builders and products have been set for Alibaba, but in the case of the products only Automate has been configured.

The credentials for Alibaba can only be set as environment variables so this has been set in the `_packer.rb` file if a credential for Alibaba exists in the databag.

The format for the databag item for Alibaba is:

```json
{
    "id": "publishing_credentials",
    "alibaba": {
        "access_key": "abcdefghijklm",
        "secret_key": "abcdefghijklm"
    }
}
```

I have created an image in Alibaba that has been created with this and the omnibus package.

russellseymour committed:
Added support for creating a marketplace image in Alibaba

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

jjasghar commented on attributes/alibaba.rb:
```diff
@@ -0,0 +1,53 @@
+  'support_email' => 'aws@chef.io',
+  'sales_email' => 'awesome@chef.io',
+  'reporting_cron_enabled' => false,
+  'doc_url' => 'https://docs.chef.io/aws_marketplace.html',
```
Should we create an Alibaba site for this?

jjasghar commented:
How is the licence done here? Is it pulled in from the cookbook? Does it default to the 30 day Automate license?

russellseymour committed:
Fixed issue with packer_provisioner template
Added URLs to download products to attributes. A new packer file is created so that the Omnibus marketplace recipe reads it in to get the URLs for download

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

russellseymour commented:
@jjasghar Not sure about that at the moment. I _think_ it will default to the 30 day license.

Still trying to work out how all of this hangs together.

russellseymour commented:
Added new `packer_provisioner` when building an image on the Alibaba platform that writes out a JSON file of the product URLs so that the `omnibus_marketplace` is able to download from OSS.

russellseymour committed:
Updated URL for chef-marketplace

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

ryancragun commented on .kitchen.yml:
```diff
@@ -16,7 +16,20 @@ platforms:
+      packman:
+        version: 1.0.4
+        checksums:
+          "1.0.4": 646da085cbcb8c666474d500a44d933df533cf4f1ff286193d67b51372c3c59e
```
While I think it's good that we're upgrading packer, I'd prefer to see the default version updated to 1.0.4 and breaking changes in other packer definitions fixed. We're tracking that work on: [CLOUD-261](https://chefio.atlassian.net/browse/CLOUD-261) and [CLOUD-262](https://chefio.atlassian.net/browse/CLOUD-262)

russellseymour replied:
Agreed - did not want to assume upgrading the default version whilst I was getting the initial Alibaba work going

ryancragun commented on attributes/alibaba.rb:
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+cred_dir = ::File.expand_path(::File.join('~', '.alibaba'))
+credential_file = ::File.join(cred_dir, 'credentials')
```
I think we should probably move the alibaba credentials into the same data_bag that has the Azure, GCP, and AWS creds.

russellseymour replied:
I have used the same format as the data bag so it should just be a case of copying and paste from what I have. I do not know where the data bag for those is to update.

To get this working I used a local version of the data bag that kitchen could read.

ryancragun commented on attributes/alibaba.rb:
```diff
@@ -0,0 +1,56 @@
+default['marketplace_image']['alibaba']['public']['automate']['enabled'] = false
+
+default['marketplace_image']['alibaba']['product_urls'] = {
+  'marketplace' => 'http://chef-software.oss-cn-beijing.aliyuncs.com/chef-marketplace_0.2.5%2B20170911103840.git.2.1c6a67f-1_amd64.deb',
```
I suspect that there's a firewall preventing the instance from accessing our package mirrors? Do we want to require the full URL or just the version of packages? What's the process of uploading a new package version to the Alibaba mirror? That should at least be documented.

russellseymour replied:
It is not a preventing of getting to mirrors, it is throttling. It is so slow to download that most of the time things time out.

At the moment we need the full URL to the package that needs to be downloaded. It was the simplest way I could think to get it working.

Currently it is a case of logging into the Alibaba cloud with the account in LastPass and manually uploading. There are command line tools that allow you to this but I have not investigated thoe yet.

ryancragun commented on attributes/alibaba.rb:
```diff
@@ -0,0 +1,56 @@
+  'role' => 'automate',
+  'platform' => 'alibaba',
+  'user' => 'ubuntu',
+  'support_email' => 'aws@chef.io',
```
definitely don't want to use the `aws` alias for alibaba.

russellseymour replied:
Agreed - my mistake I missed this.

Have changed it to `alibaba@chef.io` (although it will need creating on our side).

ryancragun commented on attributes/alibaba.rb:
```diff
@@ -0,0 +1,56 @@
+  'support_email' => 'aws@chef.io',
+  'sales_email' => 'awesome@chef.io',
+  'reporting_cron_enabled' => false,
+  'doc_url' => 'https://docs.chef.io/aws_marketplace.html',
```
we should probably make this specific to alibaba as well.

russellseymour replied:
Have changed this to an Alibaba specific address - again it needs creating.

ryancragun commented on attributes/alibaba.rb:
```diff
@@ -0,0 +1,56 @@
+  'sales_email' => 'awesome@chef.io',
+  'reporting_cron_enabled' => false,
+  'doc_url' => 'https://docs.chef.io/aws_marketplace.html',
+  'disable_outbound_traffic' => false,
```
We probably want to set this to true. Right now there are several tools that expect to be able to hit outside mirrors. Since it appears that isn't the case we'll want to set this to true.

russellseymour replied:
As mentioned its not that we cannot get to external mirrors, they are just very slow.

When using the Alibaba mirrored Ubuntu repos everything is OK, packages install as expected. This is the default in the images.

ryancragun commented on attributes/alibaba.rb:
```diff
@@ -0,0 +1,56 @@
+  'reporting_cron_enabled' => false,
+  'doc_url' => 'https://docs.chef.io/aws_marketplace.html',
+  'disable_outbound_traffic' => false,
+  'license_type' => 'flexible',
```
This should be `fixed` since we're not doing metered billing.

russellseymour replied:
Modified in latest changes

russellseymour committed:
Updated URL for chef-marketplace

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

russellseymour committed:
Changed support email address.

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

russellseymour committed:
Modified Doc URL

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

russellseymour committed:
Changed to fixed licence type

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

russellseymour committed:
Added script to ensure that the ipv6 loopback is enabled
Changed URL for the marketplace image

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>

russellseymour committed:
Changed instance size as the one we were using is now not available in Alibaba
Added script to perform a dist-upgrade on Ubuntu 14.04

Signed-off-by: Russell Seymour <russell.seymour@turtlesystems.co.uk>
